### DISCUSSION TITLE
Optimizing Grand Canonical Hybrid MCMD Simulations

### DISCUSSION THREAD

**Discussion by miramor** [glotzerlab/hoomd-blue#1946] [Support] [answered]
## Optimizing Grand Canonical Hybrid MCMD Simulations

Hello, 

I have a question about optimizing grand canonical hybrid MC/MD (Monte Carlo/Molecular Dynamics) for simulations. Here is an overview of my system and the challenges I am facing:

---

### System Overview
- **Molecular setup**:  
  My system consists of a single type of molecule, where each molecule is composed of multiple particles. 

- **Simulation goal**:  
  I want to perform standard MD simulations, allowing the exchange of molecules according to the grand canonical ensemble at a predefined frequency.

- **Current approach**:  
    To the best of my knowledge, HOOMD currently supports GCMC only for hard particles, so I created my own MCMD implementation. After running MD, I take a snapshot of the system and perform particle exchanges by calculating the total potential energy of the system. However, I believe this approach can be optimized, as discussed further below.

---

### Implementation
I use hoomd 4.6 with miniconda.
- **Molecule exchange**: I replace one molecule at a time (all molecules are identical). I use the following steps to do this. I only show the update of the positions of the particles, but the same applies to velocities, bonds and angles.
  
  ```python
  # Adding one molecule
  snapshot.particles.N += nr_molecule # nr_molecule is the number of particles comprising one moleulce
  snapshot.particles.position[-nr_molecule:] = new_position
  ...
  # Removing one molecule
  snapshot.position[start:stop] = snapshot.position[-nr_molecule:]
  snapshot.particles.N -= nr_molecule
  ...
  ```

    When removing molecules, I swap the coordinates of the molecule to be removed (with indices ranging from `[start:stop]`) with those of the last molecule in the system. The position of the last molecule is then automatically deleted by HOOMD when the particle count decreases.

- **Energy calculation**: Instead of computing the total energy of the system for each exchange, I would like to compute the energy difference by calculating only the energy contribution of the one molecule to be added or removed.
    - **Issue**: When attempting to compute the energy for a specific molecule (selected with `hoomd.filter.Tag`), I encounter a segmentation fault after numerous MC cycles. \
     In the documentation it says: "A particle tag is a unique identifier assigned to each particle in the simulation state. When the state is first initialized, it assigns tags 0 through N_particles to the particles in the order provided." \
    I assume that each particle's tag corresponds to its index like the indices in the `snapshot.position.particles array`. To define the tags of the molecule in question, I randomly select a starting index that is a multiple of nr_molecules up to `snapshot.particles.N`, and define a range of length `nr_molecules` from it. The values of the tags are therefore always in the range from 0 to `snapshot.particles.N`.
    I suspect this issue might arise from HOOMD's internal restructuring of particle tags when the particle count changes. Specifically, I am unsure how particle deletion affects the tags of the remaining particles.
    Ther error I am getting is
    ```bash
    # Cycle 1: 
    HOOMDArray([351, 352, 353, 354, 355, 356, 357, 358, 359])  # Last 9 elements of `data.particles.tag` retrieved using `simulation.state.cpu_local_snapshot`.
    [324, 325, 326, 327, 328, 329, 330, 331, 332]  # Example of random tags list representing the molecule used for exchange.
    5.70 5.70  # Energy difference calculated: (1) using the total potential energy and (2) using the energy contribution of the specific tags only.

    # Cycle 2: 
    HOOMDArray([342, 343, 344, 345, 346, 347, 348, 349, 350])  # Updated particle tags for this cycle's simulation.
    [171, 172, 173, 174, 175, 176, 177, 178, 179]  # New set of random tags.
    2.59 2.59  # Energy difference calculated similarly as in Cycle 1.

    ...

    # Cycle X: Encountering an error
    HOOMDArray([324, 325, 326, 327, 328, 329, 330, 331, 332])  
    [144, 145, 146, 147, 148, 149, 150, 151, 152]  
    invalid fastbin entry (free) 
    Aborted  
    ```
    
    I compute the potential energy of a single molecule identified by a randomly selected `tag_list`.
    ```python
        simulation.state.set_snapshot(snapshot)
        thermo_tag = hoomd.md.compute.ThermodynamicQuantities(filter=hoomd.filter.Tags(tag_list))
        simulation.operations.computes.append(thermo_tag)
        simulation.run(0)
        e = thermo_type.potential_energy
        simulation.operations.computes.remove(thermo_tag)
    ```
    Similarly, when I tried using `hoomd.filter.Type`—by assigning a specific label to the molecule being added or deleted—I encountered the same error. 

     - **Current workound**: To compute the total potential energy, I currently use `hoomd.filter.All` as a fallback. My implementation is as follows:

  ```python
    simulation.state.set_snapshot(snapshot)
    simulation.operations.computes.append(thermo_all)
    simulation.run(0)
    potential_energy = thermo_all.potential_energy
    simulation.operations.computes.remove(thermo_all)
    ```
    where `thermo_all` is predefined as 
     ```python
    thermo_all = hoomd.md.compute.ThermodynamicQuantities(filter=hoomd.filter.All())
    ```

    Using `filter.All`, I append and remove `thermo_all` from simulation.operations.computes to ensure the energy value updates correctly. Without this, the energy remains constant. While this method works, it feels somewhat inefficient to always compute the total energy. (For `thermo_tag`, I remove it from the simulation.operations.computes list as it is only needed once.)

### Question
To summarize, my questions are:

1. Is there a more efficient way to update the system configuration (e.g., changing the particle number) than using a snapshot, as demonstrated in my current implementation?
2. What is the most efficient way to calculate the energy difference between two system configurations during a particle exchange? How can it be implemented without causing segmentation faults, as occurs with my current optimization?

I would greatly appreciate any advice on improving the implementation or suggestions for best practices.

**Comment by joaander** [comment, marked as answer]
There are C++ API calls to remove and add individual particles but these are not exposed to Python (#1535). I also don't think that these methods correctly handle removing the bonds/angles/... connected to the particles, but it has been many years since I looked at that code. There are no methods currently implemented to add/remove whole molecules. 

`set_snapshot` triggers a full re-initialization of all particle/bond/angle/... data structures. It is therefore very expensive (10x or more the cost of evaluating the total energy of the system). This will also re-initialize the tags for every particle to match the index in the snapshot provided to `set_snapshot`. This should help explain the behavior you are seeing with invalid tags.

All the `Force` classes in HOOMD-blue are optimized to do one thing only - compute the forces needed for MD simulations with high performance. They always operate on the whole simulation state and they do not store the pairwise energy terms that you need to efficiently compute the deltaE due to the removal of a molecule. The only valid energy computed by these classes is the total energy on the system. The assignment of "per-particle" energies is arbitrary. It appears that I don't explicitly document this - but the convention that HOOMD-blue uses is to include 1/2 of the U_ij in each of the U_i and U_j terms. Therefore, your `thermo_tag` is not computing the energy sum that you need. There are terms in the needed deltaE that are included in U_k sums (where k is a particle outside your molecule but interacting with it).

Unfortunately, there is no solution to this problem in HOOMD-blue. Providing energy computations at this granularity would require a massive restructuring of the codebase. I also don't think that it would be possible to provide that level of customization with high performance from a Python API. At the same time, this level of customization is a common request and clearly needed by the community. I am working on some ideas in this direction but am not yet ready to share any details. I'll announce the project on this discussion board when it is ready to share: https://github.com/glotzerlab/hoomd-blue/discussions/categories/announcements

**Comment by miramor** [comment]
Thank you very much for your fast and informative reply! Do you have any estimate on when these improvements might be available?

**Reply by joaander** [reply]
I have no estimate at this time. The project is at a very early stage.

**Comment by miramor** [comment]
Thank you for the information. I am looking forward to it nonetheless.

**Reply by MartinGirard** [reply]
@miramor My group has tools to do various kinds of MC in hoomd; some of which are published. Can you send me an email with more details to know if I have something appropriate for what you want to do? (martin.girard @ mpip-mainz.mpg.de).
